source/GameAction.cpp

```diff
@@ -0,0 +1,228 @@
+/* GameAction.cpp
+Copyright (c) 2020 by Jonathan Steck
+
+Endless Sky is free software: you can redistribute it and/or modify it under the
+terms of the GNU General Public License as published by the Free Software
+Foundation, either version 3 of the License, or (at your option) any later version.
+
+Endless Sky is distributed in the hope that it will be useful, but WITHOUT ANY
+WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+*/
+
+#include "GameAction.h"
+
+#include "DataNode.h"
+#include "DataWriter.h"
+#include "Dialog.h"
+#include "Format.h"
+#include "GameData.h"
+#include "GameEvent.h"
+#include "PlayerInfo.h"
+#include "Random.h"
+
+#include <cstdlib>
+#include <vector>
+
+using namespace std;
+
+
+
+// Construct and Load() at the same time.
+GameAction::GameAction(const DataNode &node, const string &missionName)
+{
+	Load(node, missionName);
+}
+
+
+
+void GameAction::Load(const DataNode &node, const string &missionName)
+{
+	empty = false;
```

Comment:
What if we have just an action keyword, but no child entries below it. Shouldn't that be considered empty? (Set empty to false in LoadSingle?)

And do we even need the empty variable? Cannot we check the individual data-items to see if they contain data in the IsEmpty() check?

Reply:
Checked every potential thing in a GameAction to see if it has changed seemed clunkier to me than just seeing if the GameAction has ever been loaded.

Reply:
Yes, it is clunkier, but it also feels more correct to base the check on actual data instead of on metadata like having passed the load-event.

It's not a big deal to me though. Please feel free to keep the boolean as-is.